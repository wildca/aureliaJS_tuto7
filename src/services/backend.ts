import {Article} from '../models/Article';
import {Category} from '../models/Category';
import {Counter} from '../models/Counter';

export interface Backend {
    authenticate(host: string, username: string, password: string): Promise<string>;
    getArticles(categoryId: string): Promise<Article[]>;
    getCategories(): Promise<Category[]>;
    getCounters(): Promise<Counter[]>;
    /**
     * Returns true if the user is logged in. Returns false otherwise.
     *
     * @param allowAutoLogin If the session token is not valid any more and this is true,
     * we try to get a new token.
     */
    isLoggedIn({ allowAutoLogin }): Promise<boolean>;
    markAsRead(article: Article): Promise<void>;
    markAsUnread(article: Article): Promise<void>;
    markAsFavorite(article: Article): Promise<void>;
    unmarkAsFavorite(article: Article): Promise<void>;
}
