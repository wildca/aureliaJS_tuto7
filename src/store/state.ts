import environment from '../environment';
import {Article} from '../models/Article';
import {Category} from '../models/Category';

/**
 * Data from this interface is restored from localStorage. Please dont introduce breaking changes.
 */
export interface Auth {
    host: string;
    isLoggedIn: boolean;
    password: string;
    token: string;
    username: string;
}

export enum SelectableBackend {
    ttrss = 'TTRSS',
    test = 'test',
}

export interface Route {
    name: string;
    params: object;
}

export interface Router {
    currentRoute: Route;
    newRoute: Route | null;
}

export interface RSS {
    isFetching: boolean;
    categories: Category[];
    currentCategoryId: string;
    // If an article is in this array, instead of marking it as read, we remove it from the list
    // of displayed articles.
    articlesToDiscards: Article[];
    displayedArticles: Article[];
}

export interface State {
    authentication: Auth;
    isOnline: boolean;
    router: Router;
    rss: RSS;
    selectableBackends: SelectableBackend[];
    selectedBackend: SelectableBackend;
}

const selectableBackends: SelectableBackend[] = [
    SelectableBackend.ttrss,
];

if (environment.debug) {
    selectableBackends.push(SelectableBackend.test);
}

export const initialState: State = {
    authentication: {
        host: '',
        isLoggedIn: false,
        password: '',
        username: '',
        token: '',
    },
    isOnline: true,
    router: {
        currentRoute: {
            name: '',
            params: {},
        },
        newRoute: null,
    },
    rss: {
        isFetching: false,
        categories: [],
        currentCategoryId: '',
        articlesToDiscards: [],
        displayedArticles: [],
    },
    selectableBackends,
    // Use the first backend of the list as the default backend.
    selectedBackend: selectableBackends[0],
};
