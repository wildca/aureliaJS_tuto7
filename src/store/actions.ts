import {State} from './state';

export function doRouterNavigation(state: State, newRouteName: string, newRouteParams: any = {}) {
    const newState = {...state};
    newState.router = {...state.router};
    newState.router.newRoute = {
        name: newRouteName,
        params: newRouteParams,
    };
    return newState;
}

export function isOffline(state: State) {
    const newState = {...state};
    newState.isOnline = false;
    return newState;
}

export function isOnline(state: State) {
    const newState = {...state};
    newState.isOnline = true;
    return newState;
}

export function logout(state: State) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.authentication = {
        host: '',
        isLoggedIn: false,
        username: '',
        password: '',
        token: '',
    };
    newState.rss.articlesToDiscards = [];
    newState.rss.displayedArticles = [];
    newState.rss.categories = [];
    return doRouterNavigation(newState, 'login');
}

export function routerPerformedNavigation(
    state: State, currentRouteName: string, currentRouteParams,
) {
    const newState = {...state};
    newState.router = {...state.router};
    newState.rss = {...state.rss};
    newState.router.newRoute = null;
    newState.router.currentRoute = {
        name: currentRouteName,
        params: currentRouteParams,
    };

    if (currentRouteParams && currentRouteParams.categoryId) {
        newState.rss.currentCategoryId = currentRouteParams.categoryId.toString();
    } else {
        newState.rss.currentCategoryId = '';
    }

    return newState;
}
