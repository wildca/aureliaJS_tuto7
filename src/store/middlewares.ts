import {rehydrateFromLocalStorage} from 'aurelia-store';
import {State} from './state';

export function rehydratePartiallyFromLocalStorage(state: State, key?: string) {
    const stateFromLocalStorage: State = rehydrateFromLocalStorage(state, key);
    const newState = {...state};
    if (stateFromLocalStorage.authentication) {
        newState.authentication = stateFromLocalStorage.authentication;
    }

    return newState;
}
