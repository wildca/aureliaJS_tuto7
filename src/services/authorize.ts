import {autoinject} from 'aurelia-framework';
import * as LogManager from 'aurelia-logging';
import {Redirect} from 'aurelia-router';
import {Store} from 'aurelia-store';
import {getBackend} from '../store/rss-backends-actions';
import {SelectableBackend, State} from '../store/state';

@autoinject()
export class AuthorizeStep {
    private logger;
    private selectedBackend: SelectableBackend;

    public constructor(private store: Store<State>) {
        this.logger = LogManager.getLogger('aurss:services:authorize');
        this.store.state.subscribe((state) => {
            this.selectedBackend = state.selectedBackend;
        });
    }

    public run(navigationInstruction, next) {
        let redirectToLogin;
        // We are trying to access the login page. Simply continue.
        if (navigationInstruction.config.name === 'login') {
            redirectToLogin = next;
        } else {
            redirectToLogin = () => next.cancel(new Redirect('login'));
        }

        this.logger.debug('Running authorize step.');
        const isAuthenticationRequired = navigationInstruction
            .getAllInstructions()
            .some((i) => i.config.settings.auth);

        if (!isAuthenticationRequired) {
            this.logger.debug('Authorization not required, continuing');
            return next();
        }

        this.logger.info(`Checking authorization with backend ${this.selectedBackend}`);
        return getBackend(this.selectedBackend)
            .isLoggedIn({ allowAutoLogin: true })
            .then((isLoggedIn) => isLoggedIn ? next() : redirectToLogin())
            .catch((error) => {
                this.logger.error('An error occurred in authentication step', error.stack);
                return redirectToLogin();
            });
    }
}
