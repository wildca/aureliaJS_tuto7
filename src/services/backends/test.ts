import {autoinject} from 'aurelia-framework';
import {dispatchify} from 'aurelia-store';
import {Article} from '../../models/Article';
import {Category} from '../../models/Category';
import {Counter} from '../../models/Counter';
import * as actions from '../../store/rss-backends-actions';
import {Backend} from '../backend';

const FAKE_TOKEN = 'FAKE_TOKEN';
const createId = getIdCreator();

@autoinject()
export class TestBackend implements Backend {
    private favoriteArticles: Article[] = [{
        author: 'Favorite author',
        feedId: 0,
        feedTitle: 'Test feed',
        id: createId(),
        isRead: false,
        isFavorite: true,
        link: 'https://example.com',
        title: 'My favorite article',
        updatedAt: new Date(),
    }];
    private freshArticles: Article[] = [{
        author: 'Fresh author',
        feedId: 1,
        feedTitle: 'Test feed 2',
        id: createId(),
        isRead: false,
        isFavorite: false,
        link: 'https://example.com',
        title: 'Fresh article',
        updatedAt: new Date(),
    }];

    public constructor() {
    }

    public authenticate(): Promise<string> {
        return Promise.resolve(FAKE_TOKEN);
    }

    public getArticles(): Promise<Article[]> {
        return Promise.resolve(this.freshArticles);
    }

    public getCategories(): Promise<Category[]> {
        return Promise.resolve([]);
    }

    public getCounters(): Promise<Counter[]> {
        return Promise.resolve([]);
    }

    public isLoggedIn(): Promise<boolean> {
        return Promise.resolve(true);
    }

    public markAsFavorite(article: Article): Promise<void> {
        const newArticle = {
            ...article,
            isFavorite: true,
        };
        this.favoriteArticles.push(newArticle);
        return dispatchify(actions.markedAsFavorite)(article);
    }

    public markAsRead(article: Article): Promise<void> {
        return dispatchify(actions.markedAsRead)(article);
    }

    public markAsUnread(article: Article): Promise<void> {
        return dispatchify(actions.markedAsUnread)(article);
    }

    public unmarkAsFavorite(article: Article): Promise<void> {
        const index = this.favoriteArticles
            .map(favoriteArticle => favoriteArticle.id)
            .indexOf(article.id);
        if (index > -1) {
            this.favoriteArticles.splice(index, 1);
        }
        return dispatchify(actions.unmarkedAsFavorite)(article);
    }
}

function getIdCreator() {
    let id = 0;
    function _createId() {
        id++;
        return id;
    }
    return _createId;
}
