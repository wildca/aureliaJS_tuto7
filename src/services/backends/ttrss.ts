import {autoinject} from 'aurelia-framework';
import {I18N} from 'aurelia-i18n';
import * as LogManager from 'aurelia-logging';
import { Store } from 'aurelia-store';
import * as constants from '../../constants';
import {Article} from '../../models/Article';
import {Category} from '../../models/Category';
import {Counter} from '../../models/Counter';
import * as actions from '../../store/rss-backends-actions';
import { Auth, State } from '../../store/state';
import {sync} from '../../utils/sync';
import { Backend } from '../backend';

enum TTRssFeedIds {
    all = -4,
    archived = 0,
    labels = -2,
    published = -2,
    stared = -1,
    unread = -3,
}

enum TTRssArticleUpdateMode {
    setToFalse = 0,
    setToTrue = 1,
    toggle = 2,
}

enum TTRssUpdatableFields {
    starred = 0,
    published = 1,
    unread = 2,
    note = 3,
}

/**
 * Backend implementation for TTRSS.
 *
 * API reference: https://git.tt-rss.org/fox/tt-rss/wiki/ApiReference
 */
@autoinject()
export class TTRss implements Backend {
    private API_PATH = '/api/';
    private authenticationInfos: Auth;
    private parsedApiUrl: URL;
    private logger;

    public constructor(private i18n: I18N, private store: Store<State>) {
        this.logger = LogManager.getLogger('aurss:backend:ttrss');
        this.store.state.subscribe(
            (state) => this.authenticationInfos = state.authentication,
        );
    }

    public authenticate(host: string, username: string, password: string): Promise<string> {
        this.buildApiUrl(host);

        return this.queryApi({ password, op: 'login', user: username })
            .then((data) => data.session_id)
            .then((token) => {
                if (!token) {
                    this.logger.error('Authentication failed');
                    return Promise.reject();
                }

                this.store.dispatch(actions.authenticateSucceeded, {
                    host,
                    password,
                    token,
                    username,
                    isLoggedIn: true,
                });

                return Promise.resolve(token);
            });
    }

    public getArticles(categoryId: string): Promise<Article[]> {
        let id;
        if (categoryId === constants.DEFAULT_CATEGORY) {
            id = TTRssFeedIds.unread;
        } else {
            id = parseInt(categoryId, 10);
        }
        if (!Number.isInteger(id)) {
            return Promise.reject(new Error(`${categoryId} is not supported by TTRSS backend`));
        }

        return this.getArticlesFromFeedOrCategory(id)
            .then((articles: Article[]) => {
                this.store.dispatch(actions.receivedArticles, articles);
                return articles;
            });
    }

    private getArticlesFromFeedOrCategory(id: number) {
        return this.queryApi({
            // eslint-disable-next-line @typescript-eslint/camelcase
            feed_id: id,
            // Categories have an id bigger than 0. Ids lower than 0 are for special feeds.
            // eslint-disable-next-line @typescript-eslint/camelcase
            is_cat: id >= 0,
            op: 'getHeadlines',
            sid: this.authenticationInfos.token,
        })
            .then((data) => data.map((rawArticle): Article => ({
                author: rawArticle.author,
                feedId: rawArticle.feed_id,
                feedTitle: rawArticle.feed_title,
                id: rawArticle.id,
                isFavorite: rawArticle.marked,
                isRead: !rawArticle.unread,
                link: rawArticle.link,
                title: rawArticle.title,
                updatedAt: new Date(rawArticle.updated * 1000),
            })));
    }

    private queryApi(payload): Promise<any> {
        return fetch(this.apiUrl, {
            body: JSON.stringify(payload),
            method: 'POST',
        })
            .then((response) => response.json())
            .then((data) => data.content);
    }

    public getCategories(): Promise<Category[]> {
        const getCategories = this.queryApi({
            op: 'getCategories',
            sid: this.authenticationInfos.token,
        });

        const unreadCountsByCategoryId = {};
        const getUnreadCounts = this.getCounters()
            .then((counts) => counts.reduce(
                (countsByCategoryId, countData) => {
                    countsByCategoryId[countData.categoryId] = countData.unreadCount;
                    return countsByCategoryId;
                }, unreadCountsByCategoryId));

        return Promise.all([getCategories, getUnreadCounts])
            .then(([categoryData, countData]) => categoryData.map((rawCategory): Category => ({
                id: rawCategory.id.toString(),
                title: rawCategory.title,
                unreadCount: countData[rawCategory.id] || 0,
            })))
            // Categories with ids < 0 are special categories and we handle them manually.
            // So we exclude them from the list of categories here.
            .then((categories) => categories.filter((category) => category.id >= 0))
            .then((categories) => {
                const specialCategories: Category[] = [{
                    id: TTRssFeedIds.all.toString(),
                    title: this.i18n.tr('display-articles-from-category-title-all'),
                    unreadCount: unreadCountsByCategoryId['global-unread'] || 0,
                }, {
                    id: TTRssFeedIds.unread.toString(),
                    title: this.i18n.tr('display-articles-from-category-title-unread'),
                    unreadCount: unreadCountsByCategoryId[TTRssFeedIds.unread] || 0,
                }, {
                    id: TTRssFeedIds.stared.toString(),
                    title: this.i18n.tr('display-articles-from-category-title-favorites'),
                    unreadCount: unreadCountsByCategoryId[TTRssFeedIds.stared] || 0,
                }];

                return specialCategories.concat(categories);
            })
            .then((categories) => {
                this.store.dispatch(actions.receivedCategories, categories);
                return categories;
            });
    }

    public getCounters(): Promise<Counter[]> {
        return this.queryApi({
            op: 'getCounters',
            sid: this.authenticationInfos.token,
        })
            .then((counts) => counts
                // We can't match this count to anything, skip it.
                .filter((counterData) => counterData.id)
                .map((counterData): Counter => ({
                    categoryId: counterData.id.toString(),
                    unreadCount: counterData.counter,
                })))
            .then((counters) => {
                this.store.dispatch(actions.receivedCounters, counters);
                return counters;
            });
    }

    public isLoggedIn({ allowAutoLogin = true } = {}): Promise<boolean> {
        if (!this.authenticationInfos.token && allowAutoLogin) {
            this.logger.debug('No token but we can try autologging');
            return this.tryAutologging();
        } else if (!this.authenticationInfos.token) {
            this.logger.debug('No token and no autologging, we are not authenticated');
            return Promise.resolve(false);
        } else if (!this.authenticationInfos.host) {
            this.logger.debug('API host is unknown, we cannot do anything.');
            return Promise.resolve(false);
        }

        this.logger.debug('Checking token validity');
        return this.queryApi({
            op: 'isLoggedIn',
            sid: this.authenticationInfos.token,
        })
            .then((data) => data.status)
            .then((isLoggedIn) => {
                if (isLoggedIn) {
                    this.logger.debug('We are logged in');
                    this.store.dispatch(actions.isLoggedIn);
                    return isLoggedIn;
                } else if (allowAutoLogin) {
                    this.logger.debug('Token is not valid, attempting to auto log');
                    return this.tryAutologging();
                }

                this.logger.debug('Authentication failed');
                return false;
            });
    }

    public markAsRead(article: Article): Promise<void> {
        return this.updateArticle(
            article, TTRssUpdatableFields.unread, TTRssArticleUpdateMode.setToFalse,
        ).then(() => this.store.dispatch(actions.markedAsRead, article));
    }

    public markAsUnread(article: Article): Promise<void> {
        return this.updateArticle(
            article, TTRssUpdatableFields.unread, TTRssArticleUpdateMode.setToTrue,
        ).then(() => this.store.dispatch(actions.markedAsUnread, article));
    }

    public markAsFavorite(article: Article): Promise<void> {
        return this.updateArticle(
            article, TTRssUpdatableFields.starred, TTRssArticleUpdateMode.setToTrue,
        ).then(() => this.store.dispatch(actions.markedAsFavorite, article));
    }

    public unmarkAsFavorite(article: Article): Promise<void> {
        return this.updateArticle(
            article, TTRssUpdatableFields.starred, TTRssArticleUpdateMode.setToFalse,
        ).then(() => this.store.dispatch(actions.unmarkedAsFavorite, article));
    }

    private tryAutologging(): Promise<boolean> {
        this.logger.debug('Trying to auto log');
        return this.authenticate(
            this.authenticationInfos.host,
            this.authenticationInfos.username,
            this.authenticationInfos.password,
        ).then(() => {
            // If authentication succeeds, the chain will be resolved.
            this.logger.debug('Auto log succeeded');
            return true;
        }, () => {
            // If authentication failed, the chain will be rejected.
            this.logger.debug('Auto log failed');
            return false;
        });
    }

    private buildApiUrl(apiHost: string = '') {
        const host = apiHost || this.authenticationInfos.host;
        this.logger.debug('API URL is set to', host);
        this.parsedApiUrl = new URL(this.API_PATH, host);
    }

    private updateArticle(
        article: Article, field: TTRssUpdatableFields, mode: TTRssArticleUpdateMode,
    ) {
        return sync(
            this.apiUrl,
            {
                body: JSON.stringify({
                    // eslint-disable-next-line @typescript-eslint/camelcase
                    article_ids: article.id.toString(),
                    field,
                    mode,
                    op: 'updateArticle',
                    sid: this.authenticationInfos.token,
                }),
                method: 'POST',
            },
            constants.TTRSS_ACTIONS_STORAGE_NAME,
            { browserSendSuccess: (data: {content: {status: string}}) => data.content.status === 'OK' },
        ).then(null, () => {
            this.logger.error('Failed to update article.');
            return Promise.reject();
        });
    }

    private get apiUrl() {
        if (!this.parsedApiUrl) {
            this.buildApiUrl();
        }

        return this.parsedApiUrl.href;
    }
}
