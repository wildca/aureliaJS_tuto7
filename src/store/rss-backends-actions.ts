import {Container} from 'aurelia-framework';
import * as LogManager from 'aurelia-logging';
import {Store} from 'aurelia-store';
import * as constants from '../constants';
import {Article} from '../models/Article';
import {Category} from '../models/Category';
import {Counter} from '../models/Counter';
import {Backend} from '../services/backend';
import {TestBackend} from '../services/backends/test';
import {TTRss} from '../services/backends/ttrss';
import {doRouterNavigation} from './actions';
import {Auth, SelectableBackend, State} from './state';

const logger = LogManager.getLogger('aurss:store:actions:rss');

export function selectBackend(state: State, selectedBackend: SelectableBackend) {
    const newState = {...state};
    newState.selectedBackend = selectedBackend;
    return newState;
}

/**
 * This action is meant to be called when the authenticate method of the backend succeeds.
 * @param state
 * @param authInfos
 */

export function authenticateSucceeded(state: State, authInfos: Auth) {
    const newState = Object.assign({}, state);
    newState.authentication = authInfos;
    return doRouterNavigation(
        newState, 'display-articles-from-category', { categoryId: constants.DEFAULT_CATEGORY},
    );
}

/**
 * This action is meant to be called when the isLoggedIn method of the backend returns true.
 * @param state
 */
export function isLoggedIn(state: State) {
    const newState = {...state};
    newState.authentication = {...state.authentication};
    newState.authentication.isLoggedIn = true;
    return newState;
}

export function fetchArticles(state: State, categoryId: string) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.isFetching = true;

    getBackend(state.selectedBackend).getArticles(categoryId);

    return newState;
}

export function receivedArticles(state: State, articles: Article[]) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.isFetching = false;
    newState.rss.displayedArticles = articles;

    return newState;
}

export function fetchCategories(state: State) {
    getBackend(state.selectedBackend).getCategories();
    return state;
}

export function receivedCategories(state: State, categories: Category[]) {
    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.categories = categories;

    return newState;
}

export function fetchCounters(state: State) {
    getBackend(state.selectedBackend).getCounters();
    return state;
}

export function receivedCounters(state: State, counters: Counter[]) {
    const newState = {...state};
    newState.rss = {...state.rss};

    const categoryIdToCounters = counters.reduce((acc, counter) => {
        acc[counter.categoryId] = counter.unreadCount;
        return acc;
    }, {});

    newState.rss.categories = state.rss.categories.map((category) => ({
        ...category,
        unreadCount: categoryIdToCounters[category.id],
    }));

    return newState;
}

/**
 * Do a request to the backend to mark an article as read.
 * @param state
 * @param article
 * @param discard: If this is true, the article will be added to articlesToDiscards.
 */
export function markAsRead(state: State, article: Article, { discard = false } = {}) {
    getBackend(state.selectedBackend).markAsRead(article);
    if (!discard) {
        return state;
    }
    const newState = {...state};
    newState.rss = {...newState.rss};
    newState.rss.articlesToDiscards = [
        ...newState.rss.articlesToDiscards,
        article,
    ];

    return newState;
}

export function markAsUnread(state: State, article: Article) {
    getBackend(state.selectedBackend).markAsUnread(article);
    return state;
}

export function markAsFavorite(state: State, article: Article) {
    getBackend(state.selectedBackend).markAsFavorite(article);
    return state;
}

export function unmarkAsFavorite(state: State, article: Article) {
    getBackend(state.selectedBackend).unmarkAsFavorite(article);
    return state;
}

export function markedAsRead(state: State, article: Article) {
    fetchCounters(state);
    if (state.rss.articlesToDiscards.includes(article)) {
        return removeArticle(state, article);
    }

    return updateArticle(state, article, { isRead: true });
}

function removeArticle(state: State, article: Article) {
    if (!state.rss.displayedArticles.includes(article)) {
        return state;
    }

    const newState = {...state};
    newState.rss = {...state.rss};
    newState.rss.displayedArticles = [...newState.rss.displayedArticles];
    const indexToRemove = newState.rss.displayedArticles.indexOf(article);
    newState.rss.displayedArticles.splice(indexToRemove, 1);

    return newState;
}

function updateArticle(state: State, article: Article, patch) {
    const newState = {...state};
    newState.rss = {...newState.rss};

    newState.rss.displayedArticles = updateArticleInArray(
        article, newState.rss.displayedArticles, patch,
    );

    return newState;
}

function updateArticleInArray(article: Article, array: Article[], patch): Article[] {
    if (!array.includes(article)) {
        return array;
    }

    const newArray = [...array];
    const indexUpdatedArticle = newArray.indexOf(article);
    newArray[indexUpdatedArticle] = {
        ...newArray[indexUpdatedArticle],
        ...patch,
    };
    return newArray;
}

export function markedAsUnread(state: State, article: Article) {
    fetchCounters(state);
    return updateArticle(state, article, { isRead: false });
}

export function markedAsFavorite(state: State, article: Article) {
    return updateArticle(state, article, { isFavorite: true });
}

export function unmarkedAsFavorite(state: State, article: Article) {
    return updateArticle(state, article, { isFavorite: false });
}

export function openArticle(state: State, article: Article) {
    const tab = window.open(article.link, '_blank');
    if (tab) {
        tab.focus();
    }

    return state;
}

export function getBackend(selectedBackend: SelectableBackend): Backend {
    switch (selectedBackend) {
        case SelectableBackend.ttrss:
            return Container.instance.get(TTRss) as Backend;
        case SelectableBackend.test:
            return Container.instance.get(TestBackend) as Backend;
        default:
            logger.error('An unsupported backend was requested, return the default test backend');
            return Container.instance.get(TestBackend) as Backend;
    }
}

// This must be exported because to tests some component that uses dispatchify,
// these actions must be registered.
export function registerRssBackendsActions(store: Store<State>) {
    store.registerAction(
        selectBackend.name,
        selectBackend,
    );
    store.registerAction(
        authenticateSucceeded.name,
        authenticateSucceeded,
    );
    store.registerAction(
        isLoggedIn.name,
        isLoggedIn,
    );
    store.registerAction(
        markAsRead.name,
        markAsRead,
    );
    store.registerAction(
        markAsUnread.name,
        markAsUnread,
    );
    store.registerAction(
        markedAsRead.name,
        markedAsRead,
    );
    store.registerAction(
        markAsFavorite.name,
        markAsFavorite,
    );
    store.registerAction(
        unmarkAsFavorite.name,
        unmarkAsFavorite,
    );
    store.registerAction(
        markedAsUnread.name,
        markedAsUnread,
    );
    store.registerAction(
        markedAsFavorite.name,
        markedAsFavorite,
    );
    store.registerAction(
        unmarkedAsFavorite.name,
        unmarkedAsFavorite,
    );
    store.registerAction(
        openArticle.name,
        openArticle,
    );
    store.registerAction(
        fetchArticles.name,
        fetchArticles,
    );
    store.registerAction(
        receivedArticles.name,
        receivedArticles,
    );
    store.registerAction(
        fetchCategories.name,
        fetchCategories,
    );
    store.registerAction(
        receivedCategories.name,
        receivedCategories,
    );
    store.registerAction(
        fetchCounters.name,
        fetchCounters,
    );
    store.registerAction(
        receivedCounters.name,
        receivedCounters,
    );
}
