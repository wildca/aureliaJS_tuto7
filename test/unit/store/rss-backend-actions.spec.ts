import {Container} from 'aurelia-framework';
import {Backend} from '../../../src/services/backend';
import {TestBackend} from '../../../src/services/backends/test';
import {
    authenticateSucceeded,
    fetchArticles,
    markAsRead,
    markAsUnread, markedAsRead, markedAsUnread,
    receivedArticles,
} from '../../../src/store/rss-backends-actions';
import {State} from '../../../src/store/state';
import {createArticle, createReadArticle, createUnreadArticle} from '../../__fixtures__/articles-fixtures';
import {createAuthInfos, createInitialState} from '../../__fixtures__/state-fixtures';

describe('rss-backend-actions', () => {
    const dummyCategory = 'dummy-cat';
    let state: State;
    let rssBackend: Backend;

    beforeEach(() => {
        state = createInitialState();

        rssBackend = Container.instance.get(TestBackend);
    });

    describe('Actions that mutate the state but do not rely on the backend', () => {
        it('authenticateSucceeded should store authentication infos in the state and to navigate to the home', () => {
            const newState = authenticateSucceeded(state, createAuthInfos());

            // Check original state not mutated.
            expect(state.authentication).toMatchSnapshot();
            expect(newState).toMatchSnapshot();
        });

        it('markedAsRead', () => {
            const article = createUnreadArticle();
            state.rss.displayedArticles.push(article);

            const newState = markedAsRead(state, article);

            // Check original not mutated.
            expect(article.isRead).toBe(false);
            expect(newState.rss.displayedArticles[0]).not.toBe(article);
            // Check state is correct.
            expect(newState.rss.displayedArticles[0].isRead).toBe(true);
        });

        it('markAsRead and discard', () => {
            const article = createUnreadArticle();
            state.rss.displayedArticles.push(article);
            state.rss.articlesToDiscards.push(article);

            const newState = markedAsRead(state, article);

            // Check original not mutated.
            expect(newState.rss.displayedArticles[0]).not.toBe(article);
            // Check state is correct.
            expect(newState.rss.displayedArticles.length).toBe(0);
        });

        it('markedAsUnread', () => {
            const article = createReadArticle();
            state.rss.displayedArticles.push(article);

            const newState = markedAsUnread(state, article);

            // Check original not mutated.
            expect(article.isRead).toBe(true);
            expect(newState.rss.displayedArticles[0]).not.toBe(article);
            // Check state is correct.
            expect(newState.rss.displayedArticles[0].isRead).toBe(false);
        });

        it('receivedArticles', () => {
            const article = createArticle();

            const newState = receivedArticles(state, [article]);

            // Check original state not mutated.
            expect(state.rss.displayedArticles.length).toBe(0);
            expect(newState.rss.displayedArticles.length).toBe(1);
            expect(newState.rss.displayedArticles[0]).toBe(article);
        });
    });

    describe('Actions that rely on the RSS backend but do not change the state', () => {
        let article;

        beforeEach(() => {
            article = createArticle();
        });

        it('fetchArticles', () => {
            spyOn(rssBackend, 'getArticles');

            fetchArticles(state, dummyCategory);

            expect(rssBackend.getArticles).toHaveBeenCalledWith(dummyCategory);
        });

        it('markAsRead', () => {
            spyOn(rssBackend, 'markAsRead');

            markAsRead(state, article);

            expect(rssBackend.markAsRead).toHaveBeenCalledWith(article);
        });

        it('markAsUnread', () => {
            spyOn(rssBackend, 'markAsUnread');

            markAsUnread(state, article);

            expect(rssBackend.markAsUnread).toHaveBeenCalledWith(article);
        });
    });
});
