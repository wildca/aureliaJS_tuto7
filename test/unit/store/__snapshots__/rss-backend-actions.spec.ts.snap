// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rss-backend-actions Actions that mutate the state but do not rely on the backend authenticateSucceeded should store authentication infos in the state and to navigate to the home 1`] = `
Object {
  "host": "",
  "isLoggedIn": false,
  "password": "",
  "token": "",
  "username": "",
}
`;

exports[`rss-backend-actions Actions that mutate the state but do not rely on the backend authenticateSucceeded should store authentication infos in the state and to navigate to the home 2`] = `
Object {
  "authentication": Object {
    "host": "https://rss.example.com",
    "isLoggedIn": true,
    "password": "fake-password",
    "token": "fake-token",
    "username": "my-username",
  },
  "isOnline": true,
  "router": Object {
    "currentRoute": Object {
      "name": "",
      "params": Object {},
    },
    "newRoute": Object {
      "name": "display-articles-from-category",
      "params": Object {
        "categoryId": "unread",
      },
    },
  },
  "rss": Object {
    "articlesToDiscards": Array [],
    "categories": Array [],
    "currentCategoryId": "",
    "displayedArticles": Array [],
    "isFetching": false,
  },
  "selectableBackends": Array [
    "TTRSS",
    "test",
  ],
  "selectedBackend": "test",
}
`;
